pkg/controller/controller.go

```diff
@@ -84,8 +88,8 @@ type Controller interface {
 
 // New returns a new Controller registered with the Manager.  The Manager will ensure that shared Caches have
 // been synced before the Controller is Started.
-func New(name string, mgr manager.Manager, options Options) (Controller, error) {
-	c, err := NewUnmanaged(name, mgr, options)
+func New(name string, gvk *schema.GroupVersionKind, mgr manager.Manager, options Options) (Controller, error) {
```

Comment:
I considered adding gvk to Options instead. But to me it doesn't look like it would fit there.

Reply:
Why not?

Reply:
I don't think we should be doing this, this breaks the abstration and makes us hardcode the assumption "one controller has exactly one kubernetes type object it owns". So far, this assumpton is only in the builder and we should IMHO leave it like that. A controller is just a thing that has event sources, if they trigger it puts something in a q and eventually reconciles them.

Offhand examples where this new assumption wouldn't work out:
* Controller that reconciles many things based on a constant reconcile request (because easier than writing a controller per thing): https://github.com/openshift/hypershift/blob/eec759e5b6096ffc2ba823665bccce6e1c54f99a/control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go#L135
* Controller that reconciles based on keys that are not representing any resource that actually exists, just the abstraction at which the wokqeue is supposed to de-duplicate for us: https://github.com/openshift/ci-tools/blob/cd6d2f7d5ead0dbb72d75dc7a72062491ae17d6e/pkg/controller/promotionreconciler/prowjobreconciler/prowjobreconciler.go#L132

Reply:
The builder already has this information, so by setting up logging for this in the builder we can both get the right logger and keep the abstraction intact.

Reply:
I agree. Even having it in this func signature kind of breaks the abstraction.

My main concern was that GVK doesn't really seem to fit in the existing options. Especially I wouldn't know how we would end up with a clean semantic for this field, e.g. what happens if you set the GVK to another GVK that the one you're reconciling or as Alvaro mentioned what if you're not even reconciling a Kubernetes object (or multiple with the same reconciler).

The reason why I considered adding it was because I need the kind (or fallback "object") during a Reconcile to add a k/v pair which identifies the currently reconciled thing. It's not possible to only configure the key of a k/v pair on the logger.

What about adding a `LogObjectKey` (with default "object") to the options? 
Details:
* I would move adding the reconciler group / kind back into the builder
* The LogObjectKey would be set to the kind in the builder and then used during reconcile to log the k/v pair of the reconciled object.

Reply:
How about passing a `logConstructor func(reconcile.Key) logr.Logr` rather than a literal `log` into the controller and making it call that with the key prior to passing it on to the reconciler?

Reply:
Good idea. Tried to implement accordingly. Please take a look if that's what you had in mind.